packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -212,19 +212,33 @@ class _TextEditingValueAccumulator {
   }
 }
 
-/// A [TextInputFormatter] that prevents the insertion of characters
-/// matching (or not matching) a particular pattern.
+/// A [TextInputFormatter] that prevents the insertion of characters matching
+/// (or not matching) a particular pattern, by replacing the characters with the
+/// given [replacementString].
 ///
 /// Instances of filtered characters found in the new [TextEditingValue]s
-/// will be replaced with the [replacementString] which defaults to the empty
+/// will be replaced by the [replacementString] which defaults to the empty
 /// string.
 ///
-/// Since this formatter only removes characters from the text, it attempts to
-/// preserve the existing [TextEditingValue.selection] to values it would now
-/// fall at with the removed characters.
+/// A [FilteringTextInputFormatter] also tries to preserve the existing
+/// [TextEditingValue.selection] and the [TextEditingValue.composing] region,
+/// adjusting them accordingly if text within either of these ranges is replaced.
+///
+/// This formatter is typically used to match potentially recurring [Pattern]s
+/// in the new [TextEditingValue]. It never completely rejects the new
+/// [TextEditingValue] and falls back to the current [TextEditingValue] when the
+/// given [filterPattern] fails to match. Consider using a different
+/// [TextInputFormatter] such as:
+/// ```dart
+/// TextInputFormatter.withFunction((oldValue, newValue) => regExp.hasMatch(newValue.text) ? newValue : oldValue)
+/// ```
+/// for accepting/rejecting new input based on a predicate on the full string.
+/// As an example, [FilteringTextInputFormatter] typically shouldn't be used
+/// with [RegExp]s that contain positional matchers (`^` or `$`) since these
+/// patterns are usually meant for matching the whole string.
```

Comment:
This is much more clear to me and I feel like I know what to do in both situations thanks to the examples :+1: